File: motoman_driver/src/motoman/motomini_tool/scripts/fake_motomini_node.py
#!/usr/bin/env python3

import threading
import time

import actionlib
import rospy
from control_msgs.msg import FollowJointTrajectoryAction
from control_msgs.msg import FollowJointTrajectoryFeedback
from control_msgs.msg import FollowJointTrajectoryResult
from industrial_msgs.msg import RobotStatus
from industrial_msgs.srv import StopMotion
from industrial_msgs.srv import StopMotionResponse
from motoman_msgs.srv import ReadGroupIO
from motoman_msgs.srv import ReadGroupIOResponse
from motoman_msgs.srv import ReadMRegister
from motoman_msgs.srv import ReadMRegisterResponse
from motoman_msgs.srv import ReadSingleIO
from motoman_msgs.srv import ReadSingleIOResponse
from motoman_msgs.srv import SelectTool
from motoman_msgs.srv import SelectToolResponse
from motoman_msgs.srv import WriteGroupIO
from motoman_msgs.srv import WriteGroupIOResponse
from motoman_msgs.srv import WriteMRegister
from motoman_msgs.srv import WriteMRegisterResponse
from motoman_msgs.srv import WriteSingleIO
from motoman_msgs.srv import WriteSingleIOResponse
from sensor_msgs.msg import JointState
from std_srvs.srv import Trigger
from std_srvs.srv import TriggerResponse
from trajectory_msgs.msg import JointTrajectory


class FakeMotominiNode:
    def __init__(self):
        rospy.init_node('fake_motomini_node')

        self.joint_names = rospy.get_param('~joint_names', [
            'joint_1', 'joint_2', 'joint_3', 'joint_4', 'joint_5', 'joint_6'
        ])
        self.publish_rate = rospy.get_param('~publish_rate_hz', 50.0)

        self.num_joints = len(self.joint_names)
        self.current_positions = [0.0] * self.num_joints
        self.current_velocities = [0.0] * self.num_joints

        self.motion_possible = True
        self.in_motion = False
        self.in_error = False
        self.drives_powered = True

        self.single_io = {}
        self.group_io = {}
        self.mregisters = {}
        self.active_tool_by_group = {}

        self.state_lock = threading.Lock()

        self.joint_state_pub = rospy.Publisher('/joint_states', JointState, queue_size=10)
        self.feedback_pub = rospy.Publisher('/feedback_states', FollowJointTrajectoryFeedback, queue_size=10)
        self.status_pub = rospy.Publisher('/robot_status', RobotStatus, queue_size=10)

        rospy.Subscriber('/joint_path_command', JointTrajectory, self.joint_path_cb, queue_size=1)

        self.action_server = actionlib.SimpleActionServer(
            '/joint_trajectory_action',
            FollowJointTrajectoryAction,
            execute_cb=self.execute_trajectory_cb,
            auto_start=False
        )
        self.action_server.start()

        rospy.Service('/robot_enable', Trigger, self.handle_enable)
        rospy.Service('/robot_disable', Trigger, self.handle_disable)
        rospy.Service('/stop_motion', StopMotion, self.handle_stop_motion)

        rospy.Service('/read_single_io', ReadSingleIO, self.read_single_io_cb)
        rospy.Service('/write_single_io', WriteSingleIO, self.write_single_io_cb)
        rospy.Service('/read_group_io', ReadGroupIO, self.read_group_io_cb)
        rospy.Service('/write_group_io', WriteGroupIO, self.write_group_io_cb)
        rospy.Service('/read_mregister', ReadMRegister, self.read_mregister_cb)
        rospy.Service('/write_mregister', WriteMRegister, self.write_mregister_cb)
        rospy.Service('/select_tool', SelectTool, self.select_tool_cb)

        rospy.loginfo('Fake MotoMini node ready with simulated topics/actions/services')
        self.publish_loop()

    def publish_loop(self):
        rate = rospy.Rate(self.publish_rate)
        while not rospy.is_shutdown():
            self.publish_joint_and_feedback_states()
            self.publish_robot_status()
            rate.sleep()

    def publish_joint_and_feedback_states(self):
        now = rospy.Time.now()

        with self.state_lock:
            positions = list(self.current_positions)
            velocities = list(self.current_velocities)

        joint_state_msg = JointState()
        joint_state_msg.header.stamp = now
        joint_state_msg.name = self.joint_names
        joint_state_msg.position = positions
        joint_state_msg.velocity = velocities
        self.joint_state_pub.publish(joint_state_msg)

        feedback_msg = FollowJointTrajectoryFeedback()
        feedback_msg.header.stamp = now
        feedback_msg.joint_names = self.joint_names
        feedback_msg.actual.positions = positions
        feedback_msg.actual.velocities = velocities
        feedback_msg.desired.positions = positions
        feedback_msg.error.positions = [0.0] * self.num_joints
        self.feedback_pub.publish(feedback_msg)

    def publish_robot_status(self):
        msg = RobotStatus()
        msg.header.stamp = rospy.Time.now()
        msg.mode.val = 2
        msg.e_stopped.val = 0
        msg.drives_powered.val = 1 if self.drives_powered else 0
        msg.motion_possible.val = 1 if self.motion_possible else 0
        msg.in_motion.val = 1 if self.in_motion else 0
        msg.in_error.val = 1 if self.in_error else 0
        msg.error_code = 0
        self.status_pub.publish(msg)

    def execute_trajectory_cb(self, goal):
        if not self.drives_powered or not self.motion_possible:
            result = FollowJointTrajectoryResult()
            result.error_code = FollowJointTrajectoryResult.PATH_TOLERANCE_VIOLATED
            self.action_server.set_aborted(result, 'Robot is disabled')
            return

        points = goal.trajectory.points
        if not points:
            result = FollowJointTrajectoryResult()
            result.error_code = FollowJointTrajectoryResult.SUCCESSFUL
            self.action_server.set_succeeded(result)
            return

        index_by_name = {name: idx for idx, name in enumerate(self.joint_names)}
        mapped_indices = [index_by_name.get(name, -1) for name in goal.trajectory.joint_names]

        self.in_motion = True
        start_time = rospy.Time.now()
        total_duration = points[-1].time_from_start.to_sec()

        while (rospy.Time.now() - start_time).to_sec() < total_duration:
            if self.action_server.is_preempt_requested() or not self.in_motion:
                self.in_motion = False
                self.action_server.set_preempted()
                return
            time.sleep(0.01)

        final_point = points[-1]
        with self.state_lock:
            for pos_idx, value in enumerate(final_point.positions):
                if pos_idx >= len(mapped_indices):
                    break
                joint_idx = mapped_indices[pos_idx]
                if joint_idx >= 0:
                    self.current_positions[joint_idx] = value
                    self.current_velocities[joint_idx] = 0.0

        self.in_motion = False
        result = FollowJointTrajectoryResult()
        result.error_code = FollowJointTrajectoryResult.SUCCESSFUL
        self.action_server.set_succeeded(result)

    def joint_path_cb(self, msg):
        if not msg.points:
            return

        first_point = msg.points[0]
        index_by_name = {name: idx for idx, name in enumerate(self.joint_names)}
        mapped_indices = [index_by_name.get(name, -1) for name in msg.joint_names]

        with self.state_lock:
            for pos_idx, value in enumerate(first_point.positions):
                if pos_idx >= len(mapped_indices):
                    break
                joint_idx = mapped_indices[pos_idx]
                if joint_idx >= 0:
                    self.current_positions[joint_idx] = value

            if first_point.velocities:
                for vel_idx, value in enumerate(first_point.velocities):
                    if vel_idx >= len(mapped_indices):
                        break
                    joint_idx = mapped_indices[vel_idx]
                    if joint_idx >= 0:
                        self.current_velocities[joint_idx] = value

    def handle_enable(self, _req):
        self.drives_powered = True
        self.motion_possible = True
        return TriggerResponse(success=True, message='Robot Enabled')

    def handle_disable(self, _req):
        self.drives_powered = False
        self.motion_possible = False
        self.in_motion = False
        return TriggerResponse(success=True, message='Robot Disabled')

    def handle_stop_motion(self, _req):
        self.in_motion = False
        if self.action_server.is_active():
            self.action_server.set_preempted()
        return StopMotionResponse(code=0)

    def read_single_io_cb(self, req):
        value = self.single_io.get(req.address, 0)
        return ReadSingleIOResponse(message='Success', success=True, value=value)

    def write_single_io_cb(self, req):
        self.single_io[req.address] = req.value
        return WriteSingleIOResponse(message='Success', success=True)

    def read_group_io_cb(self, req):
        value = self.group_io.get(req.address, 0)
        return ReadGroupIOResponse(message='Success', success=True, value=value)

    def write_group_io_cb(self, req):
        self.group_io[req.address] = req.value
        return WriteGroupIOResponse(message='Success', success=True)

    def read_mregister_cb(self, req):
        value = self.mregisters.get(req.address, 0)
        return ReadMRegisterResponse(message='Success', success=True, value=value)

    def write_mregister_cb(self, req):
        self.mregisters[req.address] = req.value
        return WriteMRegisterResponse(message='Success', success=True)

    def select_tool_cb(self, req):
        self.active_tool_by_group[req.group_number] = req.tool_number
        return SelectToolResponse(message='Success', success=True)


if __name__ == '__main__':
    try:
        FakeMotominiNode()
    except rospy.ROSInterruptException:
        pass

File: motoman_driver/src/motoman/motomini_tool/src/fake_motomini_node.cpp
#include <algorithm>
#include <cstdint>
#include <map>
#include <mutex>
#include <string>
#include <vector>

#include <boost/bind.hpp>
#include <actionlib/server/simple_action_server.h>
#include <control_msgs/FollowJointTrajectoryAction.h>
#include <control_msgs/FollowJointTrajectoryFeedback.h>
#include <control_msgs/FollowJointTrajectoryResult.h>
#include <industrial_msgs/RobotStatus.h>
#include <industrial_msgs/StopMotion.h>
#include <motoman_msgs/ReadGroupIO.h>
#include <motoman_msgs/ReadMRegister.h>
#include <motoman_msgs/ReadSingleIO.h>
#include <motoman_msgs/SelectTool.h>
#include <motoman_msgs/WriteGroupIO.h>
#include <motoman_msgs/WriteMRegister.h>
#include <motoman_msgs/WriteSingleIO.h>
#include <ros/ros.h>
#include <sensor_msgs/JointState.h>
#include <std_srvs/Trigger.h>
#include <trajectory_msgs/JointTrajectory.h>

class FakeMotominiNode
{
public:
  FakeMotominiNode()
    : action_server_(nh_, "/joint_trajectory_action", false)
  {
    ros::NodeHandle pnh("~");
    pnh.param("publish_rate_hz", publish_rate_hz_, 50.0);

    if (!pnh.getParam("joint_names", joint_names_) || joint_names_.empty())
    {
      joint_names_ = {"joint_1", "joint_2", "joint_3", "joint_4", "joint_5", "joint_6"};
    }

    current_positions_.assign(joint_names_.size(), 0.0);
    current_velocities_.assign(joint_names_.size(), 0.0);

    joint_state_pub_ = nh_.advertise<sensor_msgs::JointState>("/joint_states", 10);
    feedback_pub_ = nh_.advertise<control_msgs::FollowJointTrajectoryFeedback>("/feedback_states", 10);
    status_pub_ = nh_.advertise<industrial_msgs::RobotStatus>("/robot_status", 10);

    joint_path_sub_ = nh_.subscribe("/joint_path_command", 1, &FakeMotominiNode::jointPathCb, this);

    robot_enable_srv_ = nh_.advertiseService("/robot_enable", &FakeMotominiNode::robotEnableCb, this);
    robot_disable_srv_ = nh_.advertiseService("/robot_disable", &FakeMotominiNode::robotDisableCb, this);
    stop_motion_srv_ = nh_.advertiseService("/stop_motion", &FakeMotominiNode::stopMotionCb, this);

    read_single_io_srv_ = nh_.advertiseService("/read_single_io", &FakeMotominiNode::readSingleIoCb, this);
    write_single_io_srv_ = nh_.advertiseService("/write_single_io", &FakeMotominiNode::writeSingleIoCb, this);
    read_group_io_srv_ = nh_.advertiseService("/read_group_io", &FakeMotominiNode::readGroupIoCb, this);
    write_group_io_srv_ = nh_.advertiseService("/write_group_io", &FakeMotominiNode::writeGroupIoCb, this);
    read_mregister_srv_ = nh_.advertiseService("/read_mregister", &FakeMotominiNode::readMRegisterCb, this);
    write_mregister_srv_ = nh_.advertiseService("/write_mregister", &FakeMotominiNode::writeMRegisterCb, this);
    select_tool_srv_ = nh_.advertiseService("/select_tool", &FakeMotominiNode::selectToolCb, this);

    action_server_.registerGoalCallback(boost::bind(&FakeMotominiNode::goalCb, this));
    action_server_.registerPreemptCallback(boost::bind(&FakeMotominiNode::preemptCb, this));
    action_server_.start();

    timer_ = nh_.createTimer(ros::Duration(1.0 / publish_rate_hz_), &FakeMotominiNode::timerCb, this);

    ROS_INFO("Fake MotoMini node ready with simulated topics/actions/services");
  }

private:
  void timerCb(const ros::TimerEvent&)
  {
    publishJointAndFeedback();
    publishRobotStatus();
  }

  void publishJointAndFeedback()
  {
    sensor_msgs::JointState joint_msg;
    control_msgs::FollowJointTrajectoryFeedback fb_msg;

    std::lock_guard<std::mutex> lock(state_mutex_);
    const ros::Time now = ros::Time::now();

    joint_msg.header.stamp = now;
    joint_msg.name = joint_names_;
    joint_msg.position = current_positions_;
    joint_msg.velocity = current_velocities_;
    joint_state_pub_.publish(joint_msg);

    fb_msg.header.stamp = now;
    fb_msg.joint_names = joint_names_;
    fb_msg.actual.positions = current_positions_;
    fb_msg.actual.velocities = current_velocities_;
    fb_msg.desired.positions = current_positions_;
    fb_msg.error.positions.assign(joint_names_.size(), 0.0);
    feedback_pub_.publish(fb_msg);
  }

  void publishRobotStatus()
  {
    industrial_msgs::RobotStatus status;
    status.header.stamp = ros::Time::now();
    status.mode.val = 2;
    status.e_stopped.val = 0;
    status.drives_powered.val = drives_powered_ ? 1 : 0;
    status.motion_possible.val = motion_possible_ ? 1 : 0;
    status.in_motion.val = in_motion_ ? 1 : 0;
    status.in_error.val = in_error_ ? 1 : 0;
    status.error_code = 0;
    status_pub_.publish(status);
  }

  void goalCb()
  {
    auto goal = action_server_.acceptNewGoal();

    if (!drives_powered_ || !motion_possible_)
    {
      control_msgs::FollowJointTrajectoryResult result;
      result.error_code = control_msgs::FollowJointTrajectoryResult::PATH_TOLERANCE_VIOLATED;
      action_server_.setAborted(result, "Robot is disabled");
      return;
    }

    if (goal->trajectory.points.empty())
    {
      control_msgs::FollowJointTrajectoryResult result;
      result.error_code = control_msgs::FollowJointTrajectoryResult::SUCCESSFUL;
      action_server_.setSucceeded(result);
      return;
    }

    const trajectory_msgs::JointTrajectoryPoint& final_point = goal->trajectory.points.back();
    std::vector<int> mapping;
    mapping.reserve(goal->trajectory.joint_names.size());
    for (const std::string& name : goal->trajectory.joint_names)
    {
      auto it = std::find(joint_names_.begin(), joint_names_.end(), name);
      if (it == joint_names_.end())
      {
        mapping.push_back(-1);
      }
      else
      {
        mapping.push_back(static_cast<int>(std::distance(joint_names_.begin(), it)));
      }
    }

    in_motion_ = true;
    const ros::Time started = ros::Time::now();
    const double duration = final_point.time_from_start.toSec();
    ros::Rate wait_rate(100.0);

    while (ros::ok() && (ros::Time::now() - started).toSec() < duration)
    {
      if (action_server_.isPreemptRequested() || !in_motion_)
      {
        in_motion_ = false;
        action_server_.setPreempted();
        return;
      }
      wait_rate.sleep();
    }

    {
      std::lock_guard<std::mutex> lock(state_mutex_);
      const size_t count = std::min(final_point.positions.size(), mapping.size());
      for (size_t i = 0; i < count; ++i)
      {
        const int mapped_idx = mapping[i];
        if (mapped_idx >= 0)
        {
          current_positions_[mapped_idx] = final_point.positions[i];
          current_velocities_[mapped_idx] = 0.0;
        }
      }
    }

    in_motion_ = false;
    control_msgs::FollowJointTrajectoryResult result;
    result.error_code = control_msgs::FollowJointTrajectoryResult::SUCCESSFUL;
    action_server_.setSucceeded(result);
  }

  void preemptCb()
  {
    in_motion_ = false;
    action_server_.setPreempted();
  }

  void jointPathCb(const trajectory_msgs::JointTrajectory::ConstPtr& msg)
  {
    if (msg->points.empty())
    {
      return;
    }

    const trajectory_msgs::JointTrajectoryPoint& pt = msg->points.front();

    std::vector<int> mapping;
    mapping.reserve(msg->joint_names.size());
    for (const std::string& name : msg->joint_names)
    {
      auto it = std::find(joint_names_.begin(), joint_names_.end(), name);
      if (it == joint_names_.end())
      {
        mapping.push_back(-1);
      }
      else
      {
        mapping.push_back(static_cast<int>(std::distance(joint_names_.begin(), it)));
      }
    }

    std::lock_guard<std::mutex> lock(state_mutex_);
    const size_t pos_count = std::min(pt.positions.size(), mapping.size());
    for (size_t i = 0; i < pos_count; ++i)
    {
      const int mapped_idx = mapping[i];
      if (mapped_idx >= 0)
      {
        current_positions_[mapped_idx] = pt.positions[i];
      }
    }

    const size_t vel_count = std::min(pt.velocities.size(), mapping.size());
    for (size_t i = 0; i < vel_count; ++i)
    {
      const int mapped_idx = mapping[i];
      if (mapped_idx >= 0)
      {
        current_velocities_[mapped_idx] = pt.velocities[i];
      }
    }
  }

  bool robotEnableCb(std_srvs::Trigger::Request&, std_srvs::Trigger::Response& res)
  {
    drives_powered_ = true;
    motion_possible_ = true;
    res.success = true;
    res.message = "Robot Enabled";
    return true;
  }

  bool robotDisableCb(std_srvs::Trigger::Request&, std_srvs::Trigger::Response& res)
  {
    drives_powered_ = false;
    motion_possible_ = false;
    in_motion_ = false;
    res.success = true;
    res.message = "Robot Disabled";
    return true;
  }

  bool stopMotionCb(industrial_msgs::StopMotion::Request&, industrial_msgs::StopMotion::Response& res)
  {
    in_motion_ = false;
    if (action_server_.isActive())
    {
      action_server_.setPreempted();
    }
    res.code.val = 0;
    return true;
  }

  bool readSingleIoCb(motoman_msgs::ReadSingleIO::Request& req, motoman_msgs::ReadSingleIO::Response& res)
  {
    const auto it = single_io_.find(req.address);
    res.value = (it == single_io_.end()) ? 0 : it->second;
    res.success = true;
    res.message = "Success";
    return true;
  }

  bool writeSingleIoCb(motoman_msgs::WriteSingleIO::Request& req, motoman_msgs::WriteSingleIO::Response& res)
  {
    single_io_[req.address] = req.value;
    res.success = true;
    res.message = "Success";
    return true;
  }

  bool readGroupIoCb(motoman_msgs::ReadGroupIO::Request& req, motoman_msgs::ReadGroupIO::Response& res)
  {
    const auto it = group_io_.find(req.address);
    res.value = (it == group_io_.end()) ? 0 : it->second;
    res.success = true;
    res.message = "Success";
    return true;
  }

  bool writeGroupIoCb(motoman_msgs::WriteGroupIO::Request& req, motoman_msgs::WriteGroupIO::Response& res)
  {
    group_io_[req.address] = req.value;
    res.success = true;
    res.message = "Success";
    return true;
  }

  bool readMRegisterCb(motoman_msgs::ReadMRegister::Request& req, motoman_msgs::ReadMRegister::Response& res)
  {
    const auto it = mregisters_.find(req.address);
    res.value = (it == mregisters_.end()) ? 0 : it->second;
    res.success = true;
    res.message = "Success";
    return true;
  }

  bool writeMRegisterCb(motoman_msgs::WriteMRegister::Request& req, motoman_msgs::WriteMRegister::Response& res)
  {
    mregisters_[req.address] = req.value;
    res.success = true;
    res.message = "Success";
    return true;
  }

  bool selectToolCb(motoman_msgs::SelectTool::Request& req, motoman_msgs::SelectTool::Response& res)
  {
    active_tool_by_group_[req.group_number] = req.tool_number;
    res.success = true;
    res.message = "Success";
    return true;
  }

  ros::NodeHandle nh_;
  ros::Timer timer_;

  ros::Publisher joint_state_pub_;
  ros::Publisher feedback_pub_;
  ros::Publisher status_pub_;
  ros::Subscriber joint_path_sub_;

  actionlib::SimpleActionServer<control_msgs::FollowJointTrajectoryAction> action_server_;

  ros::ServiceServer robot_enable_srv_;
  ros::ServiceServer robot_disable_srv_;
  ros::ServiceServer stop_motion_srv_;

  ros::ServiceServer read_single_io_srv_;
  ros::ServiceServer write_single_io_srv_;
  ros::ServiceServer read_group_io_srv_;
  ros::ServiceServer write_group_io_srv_;
  ros::ServiceServer read_mregister_srv_;
  ros::ServiceServer write_mregister_srv_;
  ros::ServiceServer select_tool_srv_;

  std::mutex state_mutex_;

  std::vector<std::string> joint_names_;
  std::vector<double> current_positions_;
  std::vector<double> current_velocities_;

  bool motion_possible_{true};
  bool in_motion_{false};
  bool in_error_{false};
  bool drives_powered_{true};

  double publish_rate_hz_{50.0};

  std::map<uint32_t, int32_t> single_io_;
  std::map<uint32_t, uint8_t> group_io_;
  std::map<uint32_t, uint16_t> mregisters_;
  std::map<uint32_t, uint32_t> active_tool_by_group_;
};

int main(int argc, char** argv)
{
  ros::init(argc, argv, "fake_motomini_node");
  FakeMotominiNode node;
  ros::spin();
  return 0;
}